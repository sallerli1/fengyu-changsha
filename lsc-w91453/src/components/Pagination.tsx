import { defineComponent, toRefs, unref, Ref, ref } from "vue";
import { type PaginationProps, paginationProps } from "./types";
import { pagination } from './index';
import './style.css'
export default defineComponent({
    name: "Pagination",
    props: paginationProps,
    setup(props: PaginationProps, { attrs, emit, slots }) {
        const jumpNumber = ref(null);
        const { currentPage: propsCurrentPage, pageSize: propsPageSize, total } = toRefs(props);
        const {
            currentPage,
            pageList,
            prev,
            next,
            pageJump
        } = pagination(propsCurrentPage, propsPageSize, total, emit);
        
        return () => (
            <ul class="pagination">
                <li>共{ total?.value }条</li>
                <li class="li-pre" onClick={prev}>上一页</li>
                {
                    pageList.map(item => {
                        return <li
                            class={`li-item ${currentPage.value === item ? 'li-item_active' : ''}`}
                            onClick={pageJump.bind(this, item)}
                        >{item}</li>
                    })
                }
                <li class="li-next" onClick={next}>下一页</li>
                <li class="li-item">
                    <span>跳转至：</span>
                    <input class="input-number" type="number" v-model={jumpNumber.value} onKeydown={pageJump.bind(this, jumpNumber.value)} />
                </li>
            </ul>
        )
    }
})