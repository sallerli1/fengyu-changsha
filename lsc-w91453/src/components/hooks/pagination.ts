import { ref, Ref, unref, watch, computed } from 'vue';

const max = 7;

export function usePagination(
    currNum: Ref<number | undefined> | undefined,
    size: Ref<number | undefined> | undefined,
    totalNum: Ref<number>,
    emit: any
) {
    const curr = unref(currNum) ?? 1;
    const _pageSize = ref(10);
    // 获取分页大小
    const pageSize = computed(() => {
        return unref(size) ?? _pageSize.value;
    });
    // 获取总分页数
    const num = computed(() => {
        return unref(Math.ceil(unref(totalNum) / unref(pageSize)))
    });
    const currentPage = ref(1);
    const total = unref(num);
    let list = [];
    // 生成页码数据
    const getPageList = () => {
        list = [];
        if (total < max) {
            for (var r = 1; r < total + 1; r++) {
                list.push(r);
            }
        } else {
            if (curr < 4) {
                list = [1, 2, 3, 4, '...', total];
            } else {
                if (curr > total - 3) {
                    list = [1, '...', total - 3, total - 2, total - 1, total];
                } else {
                    list = [1, '...', curr - 1, curr, curr + 1, '...', total];
                }
            }
        }
        return list;
    }

    // 上一页
    const prev = () => {
        if (currentPage.value === 1) {
            return;
        }
        currentPage.value--
    }

    // 下一页
    const next = () => {
        if (currentPage.value === total) {
            return;
        }
        currentPage.value++
    }

    // 跳转至某一页
    const pageJump = (num: number | string) => {
        if (typeof num === 'number') {
            currentPage.value = num
        }
    }

    // 页码发生变化时触发
    const pageChange = (page: Ref<number | string>) => {
        emit('pageChange', page.value);
    }

    // 监听分页变化
    watch(currentPage, () =>
        pageChange(currentPage)
    )
    return {
        currentPage,
        pageList: getPageList(),
        prev,
        next,
        pageJump,
        pageChange
    }
}
