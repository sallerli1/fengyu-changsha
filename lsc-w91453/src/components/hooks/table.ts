import { ref, unref, watch, computed } from 'vue';

export function useTable (data: any, type: any) {
    const showData = computed(() =>{
        const list = unref(data)
        return list
    })

    const compare = (key: string) => {
        return function(a: any, b: any){
            return type.value === 1 ? a[key] - b[key] : b[key] - a[key];	
        }
    }
    const handleSort = (item: string) => {
        if (type.value === 2) {
            type.value = 0;
            data.value = JSON.parse(JSON.stringify(showData.value));
            return;
        }
        type.value++;
        data.value.sort(compare(item));
    }
    return {
        showData,
        handleSort
    }
}

