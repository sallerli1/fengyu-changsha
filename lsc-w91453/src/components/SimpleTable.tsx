import { defineComponent, computed, ref, toRefs } from "vue";
import { type TableProps,  tableProps } from "./types/index";
import Pagination from './Pagination'
import { useTable } from './hooks/table';
import './style/index.css';

export default defineComponent({
  name: "SimpleTable",
  props: tableProps,
  setup(props: TableProps, { attrs, emit, slots }) {
    // 表头数据
    const { columns, data, pagination } = toRefs(props);

    // 排序方法，type：0无序 1正序 2逆序
    const type = ref(0);
    
    const { showData, handleSort } = useTable(data, type)

    // 页码发生变化时触发
    const pageChange = (page: number) => {
      emit('pageChange', page)
    }
    
    return () => {
      return (
        <div>
          <table
          cellspacing="0"
          border="1">
            <colgroup>
              {
                columns.value.map(column => <col width={column.width} />)
              }
            </colgroup>
            <thead>
              <tr>
                  {
                    columns.value.map(item => {
                      return <th 
                        class="cell" onClick={handleSort.bind(this,item.key)}
                      >
                        <span>{item.name}</span>
                        <span v-show={item.isSort} id="sort">{Number(type.value) === 0 ? ' 无序' : Number(type.value) === 1 ? ' 正序' : ' 倒序'}</span>
                      </th>
                    })
                  }
              </tr>
            </thead>
            <tbody>
              {
                showData.value.map((item: any) => {
                  return <tr>
                    {
                      columns.value.map(col => {
                        return <td>{item[col.key]}</td>
                      })
                    }
                  </tr>
                })
              }
            </tbody>
          </table>
          <Pagination {...pagination.value} onPageChange={pageChange} />
        </div>
      );
    };
  },
})
