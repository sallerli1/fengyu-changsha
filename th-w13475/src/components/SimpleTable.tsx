import { defineComponent, reactive, watch, toRefs, computed, ref } from "vue";
import { type TableProps, tableProps, type Sort } from "./types";
import { formatWidth, compare } from '../utils/tableUtils';
import './table.css'
export default defineComponent({
  name: "SimpleTable",
  props: tableProps,
  setup(props: TableProps, { attrs, emit, slots }) {
    // 创建props的响应式引用
    const { columns, dataSource, rowKey, pagination, filter } = toRefs(props);

    // 创建表格数据,排序记录数据的响应式引用
    let { tableData, sortObj }: { tableData: any[], sortObj: Sort } = reactive({
      tableData: [],
      sortObj: {
        sortIndex: '',
        sortType: ''
      }
    });
    // 用来记录传入的表格数据的深拷贝附件,原于还原乱序用的
    let data: any[] = [];
    // 监听父组件的dataSource的变化,创建新的副本数据
    watch(dataSource, () => {
      console.log('dataSource', dataSource)
      data = JSON.parse(JSON.stringify(dataSource.value));
      if (filter && filter.value) {
        data = data.filter(filter.value);
      }
      data.forEach((row: any, index: number) => {
        row.__id = rowKey?.value ? row[rowKey.value] : index
      });
      tableData = [...(data as never[])];
    }, { immediate: true })

    // 根据列的dataIndex获取排序的状态
    const getSortText = (key: string): string => {
      let sortType = ''
      if (sortObj.sortIndex === key) {
        sortType = sortObj.sortType;
      }
      switch (sortType) {
        case 'ASC': return '升序';
        case 'DESC': return '降序';
        default: return '无序'
      }
    }

    // 排序的逻辑
    const sort = (key: string): void => {
      // 先确定此时排序的逻辑
      if (sortObj.sortIndex !== key) {
        sortObj.sortIndex = key;
        sortObj.sortType = 'ASC'
      } else if (sortObj.sortType === '') {
        sortObj.sortType = 'ASC'
      } else if (sortObj.sortType === 'ASC') {
        sortObj.sortType = 'DESC'
      } else {
        sortObj.sortType = ''
      }
      // 恢复无序
      if (!sortObj.sortType) {
        tableData = [...(data as never[])];
        return;
      }
      // 继续排序
      tableData.sort((a, b) => compare(a[key], b[key], sortObj.sortType))
    }

    const defaultPagination = {
      pageSize: 10,
      pageNo: 1,
      total: 0,
      pageSizeOptions: [5, 10, 20],
      pageChange: undefined
    }
    // 分页的逻辑
    const { pageSize, pageNo, total, pageSizeOptions, pageChange } = toRefs(pagination?.value || defaultPagination);
    // 计算有多少页
    const pageLength = computed(() => {
      return Math.max(1, Math.ceil(total.value / pageSize.value));
    })
    // 触发分页跳转
    const pageJump = (pageSize: number, pageNo: number): void => {
      if (pageChange?.value) {
        pageChange.value(pageSize, pageNo);
      }
    }
    // 分页大小更改
    const pageSizeChange = (e: Event): void => {
      pageJump(Number((e?.target as HTMLSelectElement)?.value || ''), 1)
    }
    let toJumpPage = ref(1)
    return () => {
      return (
        <div>
          <table class="my_simple_table" cellspacing="0">
            {(columns?.value || []).map(item => {
              const { width, dataIndex } = item;
              const style = {
                width: formatWidth(width || '')
              }
              return <colgroup style={style} key={dataIndex}></colgroup>
            })}
            <thead>
              <tr>
                {(columns?.value || []).map(item => (
                  <th key={item.dataIndex} class="table_head">
                    {item.title}
                    <div v-show={item.isSort} class="sort_span" onClick={() => sort(item.dataIndex)}>
                      {getSortText(item.dataIndex)}
                    </div>
                  </th>
                ))}
              </tr>
            </thead>
            <tbody>
              {tableData.map((rowData: any) => {
                return (
                  <tr key={rowData.__id}>
                    {(columns?.value || []).map(item => <td>{rowData[item.dataIndex]}</td>)}
                  </tr>
                )
              })}
            </tbody>
          </table>
          {pagination && (
            <div class="table_pagination">
              <span
                class="page_span"
                onClick={() => { pageJump(pageSize.value, Math.max(1, pageNo.value - 1)) }}
              >
                上一页
              </span>
              {new Array(pageLength.value).fill(null).map((_, index) => (
                <span class={`page_span ${(index + 1) === pageNo.value ? 'page_span_active' : ''}`} onClick={() => { pageJump(pageSize.value, index + 1) }} >
                  {index + 1}
                </span>
              ))}
              <span
                class="page_span"
                onClick={() => { pageJump(pageSize.value, Math.min(pageLength.value, pageNo.value + 1)) }}
              >
                下一页
              </span>
              <select onChange={pageSizeChange}>
                {pageSizeOptions?.value?.map(item => (
                  <option value={item}>{item}</option>
                ))}
              </select>
              <input value={toJumpPage.value} onChange={e => { toJumpPage.value = Number((e?.target as HTMLInputElement)?.value) }}></input>
              <button onClick={() => pageJump(pageSize.value, toJumpPage.value)}>跳转</button>
            </div>
          )}
        </div>
      );
    };
  },
});