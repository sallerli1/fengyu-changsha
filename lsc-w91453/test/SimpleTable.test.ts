import { mount, shallowMount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";
import SimpleTable from "../src/components/SimpleTable";
import Pagination from "../src/components/Pagination";
import { usePagination } from '../src/components/hooks/pagination';
import { ref } from 'vue';

const columns = [
  {name: 'ID', key: 'id', width: '100px', isSort: true},
  {name: '姓名', key: 'name', width: '100px', isSort: true},
  {name: '说明', key: 'info', width: '300px', isSort: false},
]
// 生成指定长度的数组
const creatData = (size: number) => {
  const data = Array.from({length: size}, () => {
    const num = parseInt(String(100 * Math.random()), 10);
    return {
      id: num,
      name: '人物'+ num,
      info: '说明内容'
    }
  });
  return data
}

const list = creatData(10)

const baseInit = (type) => {
  return shallowMount(type, {
    props: {
      data: list,
      columns,
      pagination: {
        pageSize: 10,
        currentPage: 1,
        total: 42
      }
    },
  })
};

const wrapper = baseInit(SimpleTable);
describe("base test", () => {
  test("mount component", async () => {
    expect(wrapper.html()).toMatchSnapshot();
    const sortbtn = wrapper.findAll('#sort');
    await sortbtn[0].trigger('click');
    expect(wrapper.find('#sort').element.textContent).toBe(' 正序');
    await sortbtn[0].trigger('click');
    expect(wrapper.find('#sort').element.textContent).toBe(' 倒序');
    await sortbtn[0].trigger('click');
    expect(wrapper.find('#sort').element.textContent).toBe(' 无序');
    
  });

  test("pagination component", async () => {
    // 测试分页方法
    const current = ref(1);
    const size = ref(10);
    const total = ref(100);
    const {
      currentPage,
      pageList,
      prev,
      next,
      pageJump
    } = usePagination(current, size, total, null);
    expect(currentPage.value).toBe(1);
    expect(pageList.length).toBe(6);
    await prev();
    expect(currentPage.value).toBe(1);
    await next();
    expect(currentPage.value).toBe(2);
    await pageJump(3);
    expect(currentPage.value).toBe(3);
    // 测试分页显示
    const paginationCom = shallowMount(Pagination, {
      props: {
        pageSize: 10,
        currentPage: 1,
        total: 42
      }
    });
    const pages = paginationCom.findAll(".li-item");
    // // 初始化 默认第一页
    expect(paginationCom.find('#total').element.textContent).toBe('共42条');
    expect(paginationCom.find('.li-item_active').element.textContent).toBe('1');
  })
})
