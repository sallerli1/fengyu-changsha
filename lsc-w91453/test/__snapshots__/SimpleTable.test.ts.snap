// Vitest Snapshot v1

exports[`base test > mount component 1`] = `
"<div>
  <table cellspacing=\\"0\\" border=\\"1\\">
    <colgroup>
      <col width=\\"100px\\">
      <col width=\\"100px\\">
      <col width=\\"300px\\">
    </colgroup>
    <thead>
      <tr>
        <th class=\\"cell\\"><span>ID</span><span id=\\"sort\\"> 无序</span></th>
        <th class=\\"cell\\"><span>姓名</span><span id=\\"sort\\"> 无序</span></th>
        <th class=\\"cell\\"><span>说明</span><span id=\\"sort\\" style=\\"display: none;\\"> 无序</span></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>63</td>
        <td>人物63</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>92</td>
        <td>人物92</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>88</td>
        <td>人物88</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>88</td>
        <td>人物88</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>84</td>
        <td>人物84</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>54</td>
        <td>人物54</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>77</td>
        <td>人物77</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>60</td>
        <td>人物60</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>5</td>
        <td>人物5</td>
        <td>说明内容</td>
      </tr>
      <tr>
        <td>32</td>
        <td>人物32</td>
        <td>说明内容</td>
      </tr>
    </tbody>
  </table>
  <pagination-stub pagesize=\\"10\\" currentpage=\\"1\\" total=\\"42\\"></pagination-stub>
</div>"
`;
